pandas/core/indexes/base.py

```diff
@@ -4234,6 +4234,9 @@ def putmask(self, mask, value):
         values = self.values.copy()
         try:
             np.putmask(values, mask, self._convert_for_op(value))
+            if is_period_dtype(self.dtype):
```

Comment:
why don't you generally just do this? IOW why is this just for period dtype,

Reply:
Because PeriodIndex is the only one where .values casts to object.  I'm exploring a couple of options for longer-term solutions here